lib/SyTest/ApplicationService.pm

```diff
@@ -0,0 +1,91 @@
+package SyTest::ApplicationService;
+
+use strict;
+use warnings;
+
+use Future::Utils qw( repeat );
+
+sub new
+{
+   my $class = shift;
+   my ( $info, $await_http ) = @_;
+
+   my $path = $info->path;
+
+   # Map event types to ARRAYs of Futures
+   my %futures_by_type;
+
+   my $f = repeat {
+      $await_http->( qr{^\Q$path\E/transactions/\d+$}, sub { 1 },
+         timeout => 0,
+      )->then( sub {
+         my ( $request ) = @_;
+
+         # Respond immediately to AS
+         $request->respond_json( {} );
+
+         foreach my $event ( @{ $request->body_from_json->{events} } ) {
+            my $type = $event->{type};
+
+            my $queue = $futures_by_type{$type};
+
+            # Ignore any cancelled ones
+            shift @$queue while $queue and @$queue and $queue->[0]->is_cancelled;
+
+            if( $queue and my $f = shift @$queue ) {
```

Comment:
Not *strictly* a shadowing, because the outer `$f` isn't in scope until after the statement that introduced it, which is after the entire `repeat` call. But I agree it's still confusing to the human reader, so I've renamed it. dd15368